test/extensions/filters/network/thrift_proxy/filters/ratelimit/ratelimit_test.cc

```diff
@@ -0,0 +1,483 @@
+
+  EXPECT_EQ(ThriftProxy::FilterStatus::StopIteration, filter_->messageBegin(request_metadata_));
+
+  Http::HeaderMapPtr rl_headers{new Http::TestHeaderMapImpl{
+      {"x-ratelimit-limit", "1000"}, {"x-ratelimit-remaining", "0"}, {"retry-after", "33"}}};
+
+  EXPECT_CALL(filter_callbacks_, continueDecoding()).Times(0);
+
+  // TODO(zuercher): Headers are currently ignored, but sendLocalReply is the place to pass them.
+  EXPECT_CALL(filter_callbacks_, sendLocalReply(_, false));
+  EXPECT_CALL(filter_callbacks_.stream_info_,
+              setResponseFlag(StreamInfo::ResponseFlag::RateLimited));
+
+  Http::HeaderMapPtr h{new Http::TestHeaderMapImpl(*rl_headers)};
+  request_callbacks_->complete(RateLimit::LimitStatus::OverLimit, std::move(h));
+
+  EXPECT_EQ(1U,
+            cm_.thread_local_cluster_.cluster_.info_->stats_store_.counter("ratelimit.over_limit")
+                .value());
+}
+
+TEST_F(ThriftRateLimitFilterTest, LimitResponseRuntimeDisabled) {
+  SetUpTest(filter_config_);
+  InSequence s;
+
+  EXPECT_CALL(route_rate_limit_, populateDescriptors(_, _, _, _, _))
+      .WillOnce(SetArgReferee<1>(descriptor_));
+  EXPECT_CALL(*client_, limit(_, _, _, _))
+      .WillOnce(WithArgs<0>(Invoke([&](RateLimit::RequestCallbacks& callbacks) -> void {
+        request_callbacks_ = &callbacks;
+      })));
+
+  EXPECT_EQ(ThriftProxy::FilterStatus::StopIteration, filter_->messageBegin(request_metadata_));
+
+  EXPECT_CALL(runtime_.snapshot_, featureEnabled("ratelimit.thrift_filter_enforcing", 100))
+      .WillOnce(Return(false));
+  EXPECT_CALL(filter_callbacks_, continueDecoding());
+  request_callbacks_->complete(RateLimit::LimitStatus::OverLimit, nullptr);
+
+  EXPECT_EQ(1U,
+            cm_.thread_local_cluster_.cluster_.info_->stats_store_.counter("ratelimit.over_limit")
+                .value());
+}
+
+TEST_F(ThriftRateLimitFilterTest, ResetDuringCall) {
+  SetUpTest(filter_config_);
+  InSequence s;
+
+  EXPECT_CALL(route_rate_limit_, populateDescriptors(_, _, _, _, _))
+      .WillOnce(SetArgReferee<1>(descriptor_));
+  EXPECT_CALL(*client_, limit(_, _, _, _))
+      .WillOnce(WithArgs<0>(Invoke([&](RateLimit::RequestCallbacks& callbacks) -> void {
+        request_callbacks_ = &callbacks;
+      })));
+
+  EXPECT_EQ(ThriftProxy::FilterStatus::StopIteration, filter_->messageBegin(request_metadata_));
+
+  EXPECT_CALL(*client_, cancel());
+  filter_->onDestroy();
+}
+
+TEST_F(ThriftRateLimitFilterTest, RouteRateLimitDisabledForRouteKey) {
+  route_rate_limit_.disable_key_ = "test_key";
+  SetUpTest(filter_config_);
+
+  ON_CALL(runtime_.snapshot_, featureEnabled("ratelimit.test_key.thrift_filter_enabled", 100))
+      .WillByDefault(Return(false));
+
+  EXPECT_CALL(route_rate_limit_, populateDescriptors(_, _, _, _, _)).Times(0);
+  EXPECT_CALL(*client_, limit(_, _, _, _)).Times(0);
+
+  EXPECT_EQ(ThriftProxy::FilterStatus::Continue, filter_->messageBegin(request_metadata_));
+}
+
+TEST_F(ThriftRateLimitFilterTest, ConfigValueTest) {
```

Comment:
sorry if I missed it, but I don't think we have a test where we selectively pick actions by stage number, and make sure that only the actions pertaining to the stage configured in the test are used in the request's descriptors.

Reply:
Ah I see, you have the `Stages` test below.